code/notificationbuilder/src/test/java/com/adobe/marketing/mobile/notificationbuilder/internal/builders/ManualCarouselNotificationBuilderTest.kt

```diff
@@ -0,0 +1,303 @@
+            pushTemplate,
+            trackerActivityClass,
+            expandedLayout
+        )
+        val carouselItemsCount = pushTemplate.carouselItems.size
+        verify(exactly = carouselItemsCount) { anyConstructed<RemoteViews>().setTextViewText(any(), any()) }
+        verify(exactly = carouselItemsCount) { anyConstructed<RemoteViews>().setImageViewBitmap(any(), any()) }
+        verify(exactly = carouselItemsCount) { anyConstructed<RemoteViews>().setRemoteViewClickAction(context, trackerActivityClass, any(), any(), null, any()) }
+        verify(exactly = carouselItemsCount) { expandedLayout.addView(any(), any<RemoteViews>()) }
+        verify(exactly = carouselItemsCount) { expandedLayout.setDisplayedChild(any(), centerIndex) }
+    }
+
+    @Test
+    fun `test populateManualCarouselImages with null images`() {
+        val packageName = context.packageName
+        val centerIndex = 1
+        populateManualCarouselImages(
+            context,
+            pushTemplate.carouselItems,
+            packageName,
+            centerIndex,
+            pushTemplate,
+            trackerActivityClass,
+            expandedLayout
+        )
+
+        verify(exactly = 0) { anyConstructed<RemoteViews>().setTextViewText(any(), any()) }
+        verify(exactly = 0) { anyConstructed<RemoteViews>().setImageViewBitmap(any(), any()) }
+        verify(exactly = 0) { anyConstructed<RemoteViews>().setRemoteViewClickAction(context, trackerActivityClass, any(), any(), null, any()) }
+        verify(exactly = 0) { expandedLayout.addView(any(), any<RemoteViews>()) }
+        verify(exactly = 0) { expandedLayout.setDisplayedChild(any(), centerIndex) }
+    }
+
+    @Test
+    fun `test populateFilmstripCarouselImages with valid images`() {
+        val mockBitmap = mockkClass(Bitmap::class)
+        val newIndices = Triple(0, 1, 2)
+        every { getAssetCacheLocation() } answers { "assetCacheLocation" }
+        every { getCachedImage(any()) } answers { mockBitmap }
+        every { expandedLayout.setTextViewText(any(), any()) } returns Unit
+        every { expandedLayout.setImageViewBitmap(any(), any()) } just Runs
+        every {
+            expandedLayout.setRemoteViewClickAction(
+                context,
+                trackerActivityClass,
+                any(),
+                any(),
+                null,
+                any()
+            )
+        } returns Unit
+        populateFilmstripCarouselImages(
+            context,
+            pushTemplate.carouselItems,
+            newIndices,
+            pushTemplate,
+            trackerActivityClass,
+            expandedLayout
+        )
+
+        verify(exactly = 1) { expandedLayout.setTextViewText(any(), any()) }
+        verify(exactly = 3) { expandedLayout.setImageViewBitmap(any(), any()) }
+        verify(exactly = 1) {
+            expandedLayout.setRemoteViewClickAction(
+                context,
+                trackerActivityClass,
+                any(),
```

Comment:
Can you verify the uri passed here is as expected? You can also check case when uri is empty to make sure the template's actionUri is used